trinity/p2p/tip_monitor.py

```diff
@@ -0,0 +1,123 @@
+)
+from p2p.service import BaseService
+
+from trinity.protocol.common.peer import BaseChainPeer
+from trinity.protocol.eth import commands as eth_commands
+from trinity.protocol.eth.peer import ETHPeer, ETHPeerPool
+from trinity.protocol.les import commands as les_commands
+from trinity.protocol.les.peer import LESPeer, LESPeerPool
+
+AnyPeerPool = Union[ETHPeerPool, LESPeerPool]
+
+
+# TODO move to protocol/common/monitors.py
+class ChainTipMonitor(BaseService, PeerSubscriber):
+    """
+    Monitor for potential changes to the tip of the chain: a new peer or a new block
+    """
+    subscription_msg_types = {eth_commands.NewBlock}
+    _new_tip_commands: Set[Type[Command]]
+
+    def __init__(
+            self,
+            peer_pool: AnyPeerPool,
+            new_tip_commands: Set[Type[Command]] = None,
+            token: CancelToken = None) -> None:
+        super().__init__(token)
+        self._peer_pool = peer_pool
+        # There is one event for each subscriber, each one gets set any time new tip info arrives
+        self._subscriber_notices: Set[asyncio.Event] = set()
+        if new_tip_commands is None:
+            raise ValidationError("It is not optional to specify which commands indicate a new tip")
+        else:
+            self._new_tip_commands = new_tip_commands
+
+    async def wait_tip_info(self) -> AsyncIterator[BaseChainPeer]:
+        """
+        This iterator waits until there is potentially new tip information.
+        New tip information means a new peer connected or a new block arrived.
+        Then it yields the peer with the highest total difficulty.
+        It continues indefinitely, until this service is cancelled.
+        """
+        with self._subscriber() as new_tip_event:
+            while self.is_operational:
+                await self.wait(new_tip_event.wait())
+                new_tip_event.clear()
+                yield self._peer_pool.highest_td_peer
+
+    @property
+    def msg_queue_maxsize(self) -> int:
+        # This is a rather arbitrary value, but when the sync is operating normally we never see
+        # the msg queue grow past a few hundred items, so this should be a reasonable limit for
+        # now.
+        return 2000
+
+    def register_peer(self, peer: BasePeer) -> None:
+        self._notify_tip()
+
+    async def _handle_msg_loop(self) -> None:
+        while self.is_operational:
+            peer, cmd, msg = await self.wait(self.msg_queue.get())
+            if isinstance(cmd, eth_commands.NewBlock):
+                self._notify_tip()
+
+    def _notify_tip(self) -> None:
+        for new_tip_event in self._subscriber_notices:
+            new_tip_event.set()
+
+    async def _run(self) -> None:
+        self.run_task(self._handle_msg_loop())
+        with self.subscribe(self._peer_pool):
+            await self.events.cancelled.wait()
+
+    @contextmanager
+    def _subscriber(self) -> Iterator[asyncio.Event]:
+        new_tip_event = asyncio.Event()
```

Comment:
You should go look into `asyncio.Condition` as I think it *might* fit this model better.  I had a hard time grocking it's intended use case but after some experimentation it appears to handle something like this well where there is a shared resource (the tip peer), something in charge of managing it (the `ChainTipMonitor`), and many subscribers that want to be notified when it's available.

Nothing wrong with this API though, just something you might investigate.

Reply:
I can't think of why the waiters would need exclusive access to the tip peer.  If I'm understanding `asyncio.Condition` right (a big *if*), it would add some overhead for acquiring and releasing locks at every new tip, which is unnecessary for this use case.

Returning the highest TD peer is mostly just a convenience -- every waiter currently asks for the highest td peer as soon as a new tip is found, so this saves them that lookup.

Technically, though, this class would be just as effective to simply `yield None` and then the waiter can decide what to do. Do you think that would be clearer?

Reply:
So I believe the difference is that you can re-use an `asyncio.Condition` which would allow the subscriber to grab the new head info and then re-wait the same `condition` until the next new tip data is received.  But yes, I'm pretty sure you're right about added overhead with the various locks, as well as whether we are even understating how `asyncio.Condition` is supposed to be used.

I'm fine with whatever, I think the current implementation is plenty clear (enough).  Mostly, I want to figure out how you're supposed to use that damn `asyncio.Condition` API because I continue to be perplexed by it.